### PR TITLE
Upgrade Kafka version from 1.1.1 to 2.7.2

### PR DESCRIPTION
**Pull request by hbehboudi**

### CODE REVIEW

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 1]
```diff
@@ -182,55 +185,63 @@ public void testDirectoryDateTimePattern() throws Throwable {
...
         // Checking whether they are stored in parquet files correctly
-        List<SampleMessage> actual =
-                ParquetTestUtils.readParquetFiles(hdfsConfig, files, SampleMessage.class);
+        List<SampleMessage> actual = ParquetTestUtils.readParquetFiles(hdfsConfig, files, SampleMessage.class);
         assertEquals(messages.size(), actual.size());
         assertThat(actual, containsInAnyOrder(messages.toArray(new SampleMessage[0])));
     }
 
+    /**
+     * create a common builder for all unit tests
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
     * Creates a builder of Parquet writer instance using common configurations.
```

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 2]
```diff
@@ -182,55 +185,63 @@ public void testDirectoryDateTimePattern() throws Throwable {
...
+     */
+    private Builder<SampleMessage> createCommonBuilder() {
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
    private Builder<SampleMessage> createCommonParquetWriterBuilder() {
```

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 3]
```diff
@@ -182,55 +185,63 @@ public void testDirectoryDateTimePattern() throws Throwable {
...
+                .offsetTrackerPageSize(OFFSET_TRACKER_PAGE_SIZE);
+    }
+
     /**
      * Waits till given number of files are created.
      */
     private void waitForFiles(int fileCount, String parquetFileExtension) throws InterruptedException, IOException {
         while (findFiles(parquetFileExtension).size() < fileCount) {
-            Thread.sleep(1);
+            Thread.sleep(1_000);
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: 1

The previous choice was better. The test speed is very important.

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 4]
```diff
@@ -128,47 +118,60 @@ public void testMaxOpenDuration() throws Throwable {
...
+                }
             }
+
+            // Make sure all records are written.
+            while (writer.getTotalWrittenRecords() != numberOfMessagesSent) {
+                Thread.sleep(100L);
+            }
+
+            // Wait for the last file to be written.
+            waitForFiles(numberOfFilesWithMaxFileSize + 1, DEFAULT_PARQUET_FILE_EXTENSION);
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
           ```suggestion
           // Make sure that Parquet Writer has written the expected files. 
           int numMessages = 1000;
           int totalMessagesSent = 0;
             while (findFiles(DEFAULT_PARQUET_FILE_EXTENSION).size() < numberOfFilesWithMaxFileSize) {
                sendSampleMessages(numMessages);
                totalMessagesSent += numMessages;
            }
            
            // If there are some records remained, wait for them to be written too.
            boolean extraFileExists;
            if (writer.getTotalWrittenRecords() != totalMessagesSent) {
                    extraFileExists = true;
                    waitForFiles(numberOfFilesWithMaxFileSize + 1, DEFAULT_PARQUET_FILE_EXTENSION);
            }       
```
I changed it a little bit to make the steps and the reasons behind them, more clear.

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 5]
```diff
@@ -128,47 +118,60 @@ public void testMaxOpenDuration() throws Throwable {
...
         }
+
         List<LocatedFileStatus> files = findFiles(DEFAULT_PARQUET_FILE_EXTENSION);
-        files.sort(Comparator.comparing(LocatedFileStatus::getLen));
+
+        // Delete the incompletely generated file.
+        files.sort(Comparator.comparing(LocatedFileStatus::getModificationTime));
+        files.remove(files.size() - 1);
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
        // Remove the extra file so we can check the other files that should have maximum size.
        if (extraFileExists) {
           files.sort(Comparator.comparing(LocatedFileStatus::getModificationTime));
           files.remove(files.size() - 1);
        }
```

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 6]
```diff
@@ -128,47 +118,60 @@ public void testMaxOpenDuration() throws Throwable {
...
-                        .consumerConfig(consumerConfig)
-                        .maxFileOpenDurationSeconds(DEFAULT_MAX_FILE_OPEN_DURATION_SECONDS)
-                        .targetDir(targetPath)
-                        .protoClass(SampleMessage.class)
-                        .parser(SampleMessage.PARSER)
-                        .blockSize(10 * 1024)
-                        .maxFileSize(maxFileSize);
+        final int blockSize = 10 * 1024;
+        final int maxFileSize = 10 * blockSize;
+        final int messageCount = 1_000;
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Remove the messageCount from here and define it locally. (See my next comment)

**Review by bozorgzadeh — CHANGES_REQUESTED** [review]

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 2 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 1 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 3 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 4 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 5 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 6 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 7]
```diff
@@ -131,41 +131,43 @@ public void testMaxOpenDuration() throws Throwable {
...
         try (KafkaProtoParquetWriter<SampleMessage> writer = builder.build()) {
             writer.start();
 
+            // Make sure that Parquet Writer has written the expected files.
+            int numMessages = 1000;
+            int totalMessagesSent = 0;
             while (findFiles(DEFAULT_PARQUET_FILE_EXTENSION).size() < numberOfFilesWithMaxFileSize) {
-                if (writer.getTotalWrittenRecords() != numberOfMessagesSent) {
+                if (writer.getTotalWrittenRecords() != totalMessagesSent) {
                     Thread.sleep(100L);
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
                    Thread.sleep(1L);
```

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 8]
```diff
@@ -116,6 +117,8 @@ public void testMaxOpenDuration() throws Throwable {
             waitForFiles(1, parquetFileExtension);
         }
 
+        consumeAllRemainingRecords();
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Remove the local calls to this method and instead call it in tearDown()

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 9]
```diff
@@ -217,6 +224,28 @@ public void testDirectoryDateTimePattern() throws Throwable {
...
+        properties.put(KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        properties.put(VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+
+        try (KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(properties)) {
+            consumer.subscribe(Collections.singleton(TOPIC));
+            ConsumerRecords<byte[], byte[]> records;
+            do {
+                records = consumer.poll(Duration.ofMillis(1_000));
+            }
+            while (!records.isEmpty());
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java:238]: It does not commit anything!
It just polls them.

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java [hunk 10]
```diff
@@ -217,6 +224,28 @@ public void testDirectoryDateTimePattern() throws Throwable {
         assertThat(actual, containsInAnyOrder(messages.toArray(new SampleMessage[0])));
     }
 
+    /**
+     * This function commits uncommitted records at the end of the tests.
+     * It should be run at the end of the tests to make sure no record of current test appears in the next test.
```
Comment (bozorgzadeh) [inline src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]:
```suggestion
     * Consumes the remaining records of the topic and commits the last offset of partitions.
```
In Javadoc of the method describe 'what' it does. It is up to the caller to decide when to call it.

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 7 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 8 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 9 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java:238]: Done

File: src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java (hunk 10 above)
Comment (hbehboudi) [reply src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java]: Done

**Review by bozorgzadeh — APPROVED** [review]

**Pull request merged**